### ISSUE TITLE
Improving Download Performance

### ISSUE THREAD

**Issue by TripleFrequency** [TripleFrequency/AthamePlugin.Deezer#23]
Improving performance is a big deal, and I want to make this plugin as fast as possible. This "issue" is about addressing that, and *greatly* improving download speeds.

The idea of this speedup is to do it via multithreading. When the application requests a download link, it supports ranges (Accept-Ranges: bytes), and I plan to utilize this to improve performance. Initially (at least for testing), I am planning to use 2 MB download chunks,  spanned across however many threads are required. Assuming a minimal overhead for thread creation, I would estimate performance gains of at least 50%, if not much more.

**Comment by jonaaa20** [comment]
THAT'S what I want, do you know internet download manager? Yeah, that famous download manager, I can download deezer songs (encrypted of course) at max speed, but when I use any deezer downloader/ripper, well, damn slow... If you could make it better, this would be a big deal.

**Comment by ThePlayingGoblin** [comment]
Yeah, I would like to see faster speeds as well. Hopefully soon songs will download as fast or faster than tidal.

**Comment by TripleFrequency** [comment]
@jonaaa20 @ThePlayingGoblin I have pushed a new alpha version of 1.3. It adds multi-threading. Please report back with your performance gains and any issues. I note in the release that this *could* produce corrupted files, so please check everything you download.

**Comment by jonaaa20** [comment]
Just a question, this new download speed system now is enabled per-file or this is enabled for the whole queue? I mean, just like the download managers, 32 chunks per file, not 32 simultaneous downloads...

**Comment by TripleFrequency** [comment]
It is chunks per-file, 1 MB/chunk. So it varies a lot depending on the file, but for a 20.5 MB file, you should see 21 threads.

**Comment by jonaaa20** [comment]
Great, my PC is broken so I'll test  when i've got some free time, any problems I'll report here.

**Comment by bascurtiz** [comment]
@TripleFrequency 

http://www.deezer.com/en/playlist/3069652546

Started around primetime 20:30 local time (GMT +1):
Playlist '100' finished in: 14mins 40seconds with A4
1 error while downloading a certain track

Started around primetime 20:45 local time (GMT +1):
Playlist '100' finished in: 8mins 12seconds with A3

Speedtest.net results @ 20:53:
http://beta.speedtest.net/result/6859237752

**Comment by TripleFrequency** [comment]
@bascurtiz Did any of the tracks seem to take notably longer on A4, or was it all around slower for you? I may throw in a download chunk size in the settings for people to play around with. I'm personally on a gigabit connection, and I noted it was much faster.

**Comment by TripleFrequency** [comment]
There is now a new version, A4.1, which adds a setting to modify the chunk size. Please play around with this and report any findings.

**Comment by TripleFrequency** [comment]
Here are some of my findings. I think the slowdown could be related to the overhead involved with creating new threads. Also note that the test size here was quite small, so the individual results could just be coincidence.

* A3:
  * 23 Tracks
  * FLAC
  * 1:26.91 (86.91) seconds
  * 3.78 seconds/track

* A4.1:
  * 23 Tracks
  * FLAC
  * **1 MB chunks**
  * 1:38.22 (98.22) seconds
  * 4.27 seconds/track

* A4.1:
  * 23 Tracks
  * FLAC
  * **5 MB chunks**
  * 1:04.76 (64.76) seconds
  * **2.81 seconds/track**

* A4.1:
  * 23 Tracks
  * FLAC
  * **10 MB chunks**
  * 4:44.33 (284.33) seconds
  * 12.36 seconds/track

* A4.1:
  * 23 Tracks
  * FLAC
  * **50 MB chunks** (Basically the same as no chunking)
  * 0:38.41 (38.41) seconds
  * **1.67 seconds/track**

**Comment by bascurtiz** [comment]
Started test @ 22:25:
http://beta.speedtest.net/result/6859461910

Used the same playlist '100':
http://www.deezer.com/en/playlist/3069652546


- A3:
100 Tracks
FLAC (Quality fallback enabled)
**Finished in: 9mins 16secs**

1 error while downloading:
Bruk Out - Aidonia


- A4.1:
100 Tracks
FLAC (Quality fallback enabled)
**1 MB chunks
Finished in: 15mins 12secs**

4 errors while downloading:
Pick Up the Pieces - Average White Band
Rasputin (7" version) - Boney M.
A Match Made In Heaven - Architechs Uk
Intellectual Killer - Dj The Viper


- A4.1:
100 Tracks
FLAC (Quality fallback enabled)
**5 MB chunks
Finished in: 8mins 56secs**

4 errors while downloading:
Pick Up the Pieces - Average White Band
Rasputin (7" version) - Boney M.
A Match Made In Heaven - Architechs Uk
Intellectual Killer - Dj The Viper


- A4.1:
100 Tracks
FLAC (Quality fallback enabled)
**10 MB chunks
Finished in: 9mins 29secs**

6 errors while downloading:
Stayin' Alive (From "Saturday...) - Bee Gees
Pick Up the Pieces - Average White Band
A Match Made In Heaven - Architechs Uk
This House Is Not For Sale - Bon Jovi
Rasputin (7" version) - Boney M.
Intellectual Killer - Dj The Viper


- A4.1:
100 Tracks
FLAC (Quality fallback enabled)
**50 MB chunks
Finished in: 8mins 9secs**

4 errors while downloading:
Pick Up the Pieces - Average White Band
Intellectual Killer - Dj The Viper
Rasputin (7" version) - Boney M.
A Match Made In Heaven - Architechs Uk


Ended test @ 23:31:
http://beta.speedtest.net/nl/result/6859601326
